Programs/python.c

```diff
@@ -15,6 +15,110 @@ wmain(int argc, wchar_t **argv)
 }
 #else
 
+/* Helpers to better handle the legacy C locale
+ *
+ * The legacy C locale assumes ASCII as the default text encoding, which
+ * causes problems not only for the CPython runtime, but also other
+ * components like GNU readline.
+ *
+ * Accordingly, when the CLI detects it, it attempts to coerce it to a
+ * more capable UTF-8 based alternative.
+ *
+ * See the documentation of the PYTHONCOERCECLOCALE setting for more details.
+ *
+ */
+
+#ifdef PY_COERCE_C_LOCALE
+static const char *_C_LOCALE_COERCION_WARNING =
+    "Python detected LC_CTYPE=C: %.20s coerced to %.20s (set another locale "
+    "or PYTHONCOERCECLOCALE=0 to disable this locale coercion behaviour).\n";
+
+typedef struct _CandidateLocale {
+    const char *locale_name;
+    int category;
+} _LocaleCoercionTarget;
+
+static _LocaleCoercionTarget _TARGET_LOCALES[] = {
+    { "C.UTF-8", LC_ALL },
+    { "C.utf8", LC_ALL },
+    { "UTF-8", LC_CTYPE },
+    { NULL, 0 }
+};
+
+void
+_coerce_default_locale_settings(const _LocaleCoercionTarget *target)
+{
+    const char *newloc = target->locale_name;
+    int category = target->category;
+
+    /* Reset locale back to currently configured defaults */
+    setlocale(LC_ALL, "");
+
+    /* Set the relevant locale environment variables */
+    if (category == LC_ALL) {
+        const char *env_vars_updated = "LC_ALL & LANG";
+        if (setenv("LC_ALL", newloc, 1)) {
+            fprintf(stderr,
+                    "Error setting LC_ALL, skipping C locale coercion\n");
+            return;
+        }
+        if (setenv("LANG", newloc, 1)) {
+            fprintf(stderr,
+                    "Error setting LANG during C locale coercion\n");
+            env_vars_updated = "LC_ALL";
+        }
+        fprintf(stderr, _C_LOCALE_COERCION_WARNING, env_vars_updated, newloc);
+    } else if (category == LC_CTYPE) {
+        if (setenv("LC_CTYPE", newloc, 1)) {
+            fprintf(stderr,
+                    "Error setting LC_CTYPE, skipping C locale coercion\n");
+            return;
+        }
+        fprintf(stderr, _C_LOCALE_COERCION_WARNING, "LC_CTYPE", newloc);
+    } else {
+        fprintf(stderr, "Locale coercion must target LC_ALL or LC_CTYPE\n");
+        return;
+    }
+
+    /* Set PYTHONIOENCODING if not already set */
+    if (setenv("PYTHONIOENCODING", "utf-8:surrogateescape", 0)) {
+        fprintf(stderr,
+                "Error setting PYTHONIOENCODING during C locale coercion\n");
```

Comment:
This may break old Python 2 in subprocess.

```
$ cat t.py
# encoding: utf-8
print(u"こんにちは")

$ /usr/bin/python2.6 t.py
こんにちは

$ export PYTHONIOENCODING=utf-8:surrogateescape
$ /usr/bin/python2.6 t.py
TypeError: an integer is required
```